Src/VimMac/VimKeyProcessor.cs

```diff
@@ -143,13 +94,61 @@ public override void KeyDown(KeyEventArgs e)
 
             var status = Mac.StatusBar.GetStatus(VimBuffer);
             var text = status.Text;
-            if(VimBuffer.ModeKind == ModeKind.Command)
+            if (VimBuffer.ModeKind == ModeKind.Command)
             {
                 // Add a fake 'caret'
                 text = text.Insert(status.CaretPosition, "|");
             }
             IdeApp.Workbench.StatusBar.ShowMessage(text);
             e.Handled = handled;
         }
+
+        /// <summary>
+        /// Try and process the given KeyInput with the IVimBuffer.  This is overridable by 
+        /// derived classes in order for them to prevent any KeyInput from reaching the 
+        /// IVimBuffer
+        /// </summary>
+        private bool TryProcess(KeyInput keyInput)
+        {
+            return VimBuffer.CanProcessAsCommand(keyInput) && VimBuffer.Process(keyInput).IsAnyHandled;
+        }
+
+        private bool KeyEventIsDeadChar(KeyEventArgs e)
+        {
+            return string.IsNullOrEmpty(e.Characters);
+        }
+
+        private bool IsEscapeKey(KeyEventArgs e)
+        {
+            return (NSKey)e.Event.KeyCode == NSKey.Escape;
+        }
+
+        private bool ShouldBeProcessedByVim(KeyEventArgs e)
+        {
+            if (KeyEventIsDeadChar(e))
+                // When a dead key combination is pressed we will get the key down events in 
+                // sequence after the combination is complete.  The dead keys will come first
+                // and be followed the final key which produces the char.  That final key 
+                // is marked as DeadCharProcessed.
+                //
+                // All of these should be ignored.  They will produce a TextInput value which
+                // we can process in the TextInput event
+                return false;
+
+            if (_completionBroker.IsCompletionActive(TextView) && !IsEscapeKey(e))
+                return false;
+
+            if (_signatureHelpBroker.IsSignatureHelpActive(TextView))
+                return false;
+
+            if (_inlineRenameListenerFactory.InRename)
+                return false;
+
+            if (VimBuffer.Mode.ModeKind == ModeKind.Insert && e.Characters == "\t")
+                // Allow tab key to work for snippet completion
```

Comment:
Is there a way to check to see if snippet completion is active?

Reply:
I supposed that there must be a way of checking for this. The intellisense window doesn't necessarily need to be shown for this though.

I'll try and figure out how snippet completion works.

Reply:
There are 2 different cases that we need to support here. 

The first is just before the snippet session is active (when the caret is at the end of `ctor` and tab key is pressed)

The second is during an active snippet session to tab (or back tab) between placeholders.

Reply:
Does VS4Mac have an API to track when the snippet sessions are active? VS has one but it's pretty ancient. One of the old COM based APIs. Don't think that it will be in the shared layer. 

Actually on second thought I think this is all okay. This doesn't kill the key stroke, it just means it won't be processed by the vim layer. Instead it goes into VS4Mac where it gets handled.

Reply:
I figured out how to handle the first case, but there isn't a good API to handle the 2nd case (although it would be possible with a lot of reflection hacking)

I think my original code is OK too. The only time I think it would affect anybody is if they wanted to map the tab key in insert mode to do something else (seems unlikely)